belt-dwp/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+use aead::{AeadInPlaceDetached, PostfixTagged};
+use belt_block::cipher::{Block, BlockCipherEncrypt, KeyIvInit, StreamCipher};
+use belt_block::{BeltBlock, belt_block_raw};
+use belt_ctr::BeltCtr;
+use universal_hash::UniversalHash;
+
+use crate::{
+    ghash::GHash,
+    utils::{from_u32, to_u32},
+};
+
+/// Nonce type for [`BeltDwp`]
+pub type Nonce = aead::Nonce<BeltDwp>;
+
+/// Tag type for [`BeltDwp`]
+pub type Tag = aead::Tag<BeltDwp>;
+
+mod gf;
+mod ghash;
+mod utils;
+
+/// T from the STB 34.101.31-2020
+const T: u128 = 0xE45D4A588E006D363BF5080AC8BA94B1;
+
+/// Belt-DWP authenticated encryption with associated data (AEAD) cipher, defined in
+/// STB 34.101.31-2020
+pub struct BeltDwp {
+    key: Key<BeltBlock>,
+}
+
+impl KeySizeUser for BeltDwp {
+    type KeySize = U32;
+}
+
+impl AeadInPlaceDetached for BeltDwp {
+    fn encrypt_in_place_detached(
+        &self,
+        nonce: &Nonce,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+    ) -> aead::Result<Tag> {
+        Cipher::new(self.key, nonce).encrypt_in_place_detached(associated_data, buffer)
+    }
+
+    fn decrypt_in_place_detached(
+        &self,
+        nonce: &Nonce,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+        tag: &Tag,
+    ) -> aead::Result<()> {
+        Cipher::new(self.key, nonce).decrypt_in_place_detached(associated_data, buffer, tag)
+    }
+}
+
+impl PostfixTagged for BeltDwp {}
+
+struct Cipher {
+    enc_cipher: BeltCtr,
+    mac_cipher: BeltBlock,
+    ghash: GHash,
+}
+
+impl Cipher {
+    fn new(key: Key<BeltBlock>, nonce: &Nonce) -> Self {
+        let cipher: BeltCtr = BeltCtr::new(&key, nonce);
+
+        let _s = to_u32::<4>(nonce);
+        let _k = to_u32::<8>(&key);
+        // 2.1. 𝑠 ← belt-block(𝑆, 𝐾);
+        let s = belt_block_raw(_s, &_k);
+        // 2.2. 𝑟 ← belt-block(𝑠, 𝐾);
+        let r = from_u32::<16>(&belt_block_raw(s, &_k));
+
+        Self {
+            enc_cipher: cipher,
+            mac_cipher: BeltBlock::new(&key),
+            // Unwrap is safe because the key is always 16 bytes
+            ghash: GHash::new_with_init_block(&Key::<GHash>::try_from(&r[..]).unwrap(), T),
+        }
+    }
+
+    fn encrypt_in_place_detached(
+        &mut self,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+    ) -> aead::Result<Tag> {
+        let sizes_block =
+            get_sizes_block(associated_data.len() as u64 * 8, buffer.len() as u64 * 8);
+
+        // 3. For 𝑖 = 1, 2, . . . , 𝑚 do:
+        //  3.1 𝑡 ← 𝑡 ⊕ (𝐼𝑖 ‖ 0^{128−|𝐼𝑖|})
+        //  3.2 𝑡 ← 𝑡 * 𝑟.
+        self.ghash.update_padded(associated_data);
+
+        // 4. For 𝑖 = 1, 2, . . . , 𝑛 do:
+        //  4.1 𝑠 ← 𝑠 ⊞ ⟨1⟩_128
+        //  4.2 𝑌𝑖 ← 𝑋𝑖 ⊕ Lo(belt-block(𝑠, 𝐾), |𝑋𝑖|)
+        //  4.3 𝑡 ← 𝑡 ⊕ (𝑌𝑖 ‖ 0^{128−|𝑌𝑖|})
+        //  4.4 𝑡 ← 𝑡 * 𝑟.
+        buffer.chunks_mut(16).for_each(|block| {
+            self.enc_cipher.apply_keystream(block);
+            self.ghash.update_padded(block);
```

Comment:
You should use `BeltCtrCore::process_with_backend` here.

Reply:
I also don't really understand how to do this better? I should create a separate structure in which to place my blocks and implement StreamCipherBackend?

Reply:
Not quite, you need to implement the `StreamClosure` trait, which would accept `StreamCipherBackend` (effectively, you need to manually implement closure `|stream_backend| { ... }`).

I guess we can keep the current code as-is for now and migrate to `BeltCtrCore` later in a separate PR.